### DISCUSSION TITLE
How to Write Report When App Crash?

### DISCUSSION THREAD

**Discussion by KiddoV** [wailsapp/wails#2682] [Q&A]
How do I write a report to, says text (.txt) file, if my complied app (.exe) crash?

In other words, simply write all logs to a file, when app crashed.

Thanks,

**Comment by ecsdantas** [comment]
You can create a function Init to set the default output log as:
```
func Init(){
	logpath := "C:\\MyLogFile.log"
	flog, err = os.OpenFile(logpath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0777)
	log.SetFlags(3)
	log.SetOutput(flog)
}
```

After that, just use `log.Println("information")` and `log.Panic(err)` where you want.

**Comment by leaanthony** [comment]
Open a ticket for a crash reporter. It'll be a function option that receives an error. We can then write a top level recover method that will call this option if available

**Reply by KiddoV** [reply]
Any update on this... "feature"? Maybe in V3? 🤔

Or maybe something like this?.
```go
app := NewApp()

err := wails.Run(&options.App{
    [...]
    OnCrash: app.onCrash, // Call on app crash or before it crash? (I mean PANIC or FATAL)
})
```
Sorry I forgot to open the ticket, I was out of the loop, not sure if this feature added?

**Comment by RichardLindhout** [comment]
@leaanthony does this work in v3?

**Comment by RichardLindhout** [comment]
I did do this the following way now

I created a package crashdialog
```go
// crashdialog_darwin.go
//go:build darwin
// +build darwin

package crashdialog

import (
	"fmt"
	"os/exec"
)

// ShowCrashDialog displays a native dialog on macOS with the crash report inside a text input for easy copying.
func ShowCrashDialog(crashReport string) error {
	appleScript := fmt.Sprintf(`
		tell application "System Events"
			display dialog "The app has crashed. Please copy the crash report below and send it to support@webridge.nl:" default answer "%s" buttons {"OK"} default button "OK" with icon caution
		end tell`, crashReport)

	cmd := exec.Command("osascript", "-e", appleScript)
	return cmd.Run()
}
```

windows
```go
// crashdialog_windows.go
//go:build windows
// +build windows

package crashdialog

import (
	"fmt"
	"log"
	"syscall"
	"unsafe"
)

var (
	user32          = syscall.NewLazyDLL("user32.dll")
	procMessageBoxW = user32.NewProc("MessageBoxW")
)

const (
	MB_ICONERROR = 0x00000010
	MB_OK        = 0x00000000
)

// ShowCrashDialog displays a simple native MessageBox with the crash report.
func ShowCrashDialog(crashReport string) error {
	message := fmt.Sprintf("The app has crashed. Please copy the crash report below and send it to support@webridge.nl:\n\n%s", crashReport)
	title := "App Crash"

	// Windows native MessageBox (without text input but keeps it very simple)
	t, _ := syscall.UTF16PtrFromString(title)
	c, _ := syscall.UTF16PtrFromString(message)

	// Call MessageBoxW from user32.dll
	ret, _, _ := procMessageBoxW.Call(0, uintptr(unsafe.Pointer(c)), uintptr(unsafe.Pointer(t)), MB_ICONERROR|MB_OK)
	if ret == 0 {
		log.Printf("Failed to show dialog")
		return fmt.Errorf("failed to show dialog")
	}
	return nil
}

```

```go
func panicHandler(output string) {
	if err := crashdialog.ShowCrashDialog(output); err != nil {
		log.Error().Err(err).Msg("Failed to show crashdialog dialog")
	}
}

func main() {
	exitStatus, err := panicwrap.BasicWrap(panicHandler)
	if err != nil {
		// Something went wrong setting up the panic wrapper. Unlikely,
		// but possible.
		panic(err)
	}

	// If exitStatus >= 0, then we're the parent process and the panicwrap
	// re-executed ourselves and completed. Just exit with the proper status.
	if exitStatus >= 0 {
		os.Exit(exitStatus)
	}

	app := NewApp()
```

**Reply by KiddoV** [reply]
I assumed you are using `"github.com/mitchellh/panicwrap"`?

**Comment by RichardLindhout** [comment]
Hmmm. This still does silently crash without the alert. I don't know why it does..

**Comment by lysShub** [comment]
windows, set `windows.SetStdHandle` at start

**Comment by COD3HUNT3R** [comment]
Maybe Sentry(sentry.io) could be an option.

**Comment by leaanthony** [comment]
Great solutions shared here — thanks everyone for contributing workarounds!

To summarize what works today:

**For file-based logging before a crash**, the `log.SetOutput(file)` approach shared above is reliable for capturing pre-crash output.

**For panic/crash detection with a user-facing dialog**, [@RichardLindhout](https://github.com/RichardLindhout)'s `panicwrap`-based approach is the most complete community solution. The silent crash issue likely comes from the panic happening before the wrapper process has a chance to intercept it — make sure `panicwrap.BasicWrap` is the very first thing in `main()`, before any Wails or application initialization.

**v3 note:** A first-class crash reporter option (a function hook that fires on unhandled panics) has been requested — please 👍 [the feature request](https://github.com/wailsapp/wails/issues) if you want it prioritized. In the meantime the `panicwrap` approach works for both v2 and v3.

If the silent crash issue persists after putting `BasicWrap` first, please open a separate issue with a minimal reproducer — happy to dig in.
